### PR TITLE
Introduce ra_proc_macro_srv

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR add preliminary for server side of proc macro : 

1. Add crate setup
2. IO for server side

### CODE REVIEW

File: crates/ra_proc_macro_srv/Cargo.toml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+version = "0.1.0"
+authors = ["rust-analyzer developers"]
+publish = false
+
+[lib]
+doctest = false
+
+[dependencies]
+ra_tt = { path = "../ra_tt" }
+ra_mbe = { path = "../ra_mbe" }
```
Comment (edwin0cheng) [inline crates/ra_proc_macro_srv/Cargo.toml:13]: @matklad  We need mbe here for parsing string to `ra_tt::Token` which is used for lib_proc_macro bridge api.

**Comment by edwin0cheng** [comment]
Failed in `deny_c` for `libloading` crate. Um...

**Comment by Veetaha** [comment]
Maybe use `dlopen()` and `LoadLibrary()` directly?

**Comment by bjorn3** [comment]
The C code is used for a lock around `dlerror`. It can't be written in C, as it uses a weak global variable to store the lock in. This way multiple versions of `libloading` can use the same lock without conflicts.

https://github.com/nagisa/rust_libloading/blob/953681d3958c04d735b30a1170b3fa6fc8ed7ab0/src/os/unix/global_static.c

```c
#include <pthread.h>
#include <stdlib.h>

pthread_mutex_t __attribute__((weak)) rust_libloading_dlerror_mutex = PTHREAD_MUTEX_INITIALIZER;

void __attribute__((weak))
rust_libloading_dlerror_mutex_lock(void)
{
    if (pthread_mutex_lock(&rust_libloading_dlerror_mutex) != 0) {
        abort();
    }
}

void __attribute__((weak))
rust_libloading_dlerror_mutex_unlock(void)
{
    if (pthread_mutex_unlock(&rust_libloading_dlerror_mutex) != 0) {
        abort();
    }
}
```

**Comment by edwin0cheng** [comment]
Another reasons : We would need to get proc-macro symbols from these dylibs in different platform (linux, mac, windows). and using `libloading` to handle them fairly easily.

**Comment by matklad** [comment]
>Maybe use dlopen() and LoadLibrary() directly?

This is exactly the case where depending on 3d-party lib can save us a royal amount of pain :)

**Comment by bjorn3** [comment]
Maybe `deny_c` can become a script that checks if it's parent process is the build script for `libloading` and only execute the c compiler in that case?

**Comment by flodiebold** [comment]
Maybe we should have a cargo feature for enabling/disabling proc macro support, and could then do the deny_c check without that?

**Comment by edwin0cheng** [comment]
> Maybe we should have a cargo feature for enabling/disabling proc macro support, and could then do the deny_c check without that?

That deny_c check is defined in environment variable, so we would have to separate two CL check action to support it. Anyway,  we have to change our CI for the proc-macro-srv executable to be included in the release package.

**Comment by matklad** [comment]
I'm ok with just disabling deny_c check. I've also opened an upstream issue
about making the c dep optional (which I think it doable).

**Pull request merged**